cmd/kubeadm/app/cmd/upgrade/diff.go

```diff
@@ -0,0 +1,141 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upgrade
+
+import (
+	"io/ioutil"
+	"os"
+
+	"github.com/golang/glog"
+	"github.com/pmezard/go-difflib/difflib"
+	"github.com/spf13/cobra"
+	corev1 "k8s.io/api/core/v1"
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	"k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/scheme"
+	kubeadmapiext "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha2"
+	cmdutil "k8s.io/kubernetes/cmd/kubeadm/app/cmd/util"
+	"k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	"k8s.io/kubernetes/cmd/kubeadm/app/phases/controlplane"
+	kubeadmutil "k8s.io/kubernetes/cmd/kubeadm/app/util"
+	configutil "k8s.io/kubernetes/cmd/kubeadm/app/util/config"
+	"k8s.io/kubernetes/pkg/util/version"
+)
+
+type diffFlags struct {
+	apiServerManifestPath         string
+	controllerManagerManifestPath string
+	schedulerManifestPath         string
+	newK8sVersionStr              string
+	contextLines                  int
+	parent                        *cmdUpgradeFlags
+}
+
+var (
+	defaultAPIServerManifestPath         = constants.GetStaticPodFilepath(constants.KubeAPIServer, constants.GetStaticPodDirectory())
+	defaultControllerManagerManifestPath = constants.GetStaticPodFilepath(constants.KubeControllerManager, constants.GetStaticPodDirectory())
+	defaultSchedulerManifestPath         = constants.GetStaticPodFilepath(constants.KubeScheduler, constants.GetStaticPodDirectory())
+)
+
+// NewCmdDiff returns the cobra command for `kubeadm upgrade diff`
+func NewCmdDiff(parentflags *cmdUpgradeFlags) *cobra.Command {
+	flags := &diffFlags{
+		parent: parentflags,
+	}
+
+	cmd := &cobra.Command{
+		Use:   "diff [version]",
+		Short: "Show what differences would be applied to existing static pod manifests",
+		Run: func(cmd *cobra.Command, args []string) {
+			runDiff(flags, args)
+		},
+	}
+
+	cmd.Flags().StringVar(&flags.apiServerManifestPath, "api-server-manifest", defaultAPIServerManifestPath, "path to API server manifest")
+	cmd.Flags().StringVar(&flags.controllerManagerManifestPath, "controller-manager-manifest", defaultControllerManagerManifestPath, "path to controller manifest")
+	cmd.Flags().StringVar(&flags.schedulerManifestPath, "scheduler-manifest", defaultSchedulerManifestPath, "path to scheduler manifest")
+	cmd.Flags().IntVarP(&flags.contextLines, "context-lines", "c", 3, "How many lines of context in the diff")
+
+	return cmd
+}
+
+func runDiff(flags *diffFlags, args []string) {
+	var cfg *kubeadmapi.MasterConfiguration
+
+	// If the version is specified in config file, pick up that value.
+	if flags.parent.cfgPath != "" {
```

Comment:
Technically I don't think you should need to do this, as `ConfigFileAndDefaultsToInternalConfig` has this conditional inside of itself.